src/detect-engine-alert.c

```diff
@@ -275,6 +203,105 @@ static void PacketApplySignatureActions(Packet *p, const Signature *s, const uin
+    void *tmp_queue = SCRealloc(det_ctx->alert_queue, (size_t)(sizeof(PacketAlert) * new_cap));
+    if (unlikely(tmp_queue == NULL)) {
+        /* if first realloc fail, try to grow just 50% */
+        new_cap = (uint16_t)(det_ctx->alert_queue_capacity * 3 / 2);
+        tmp_queue = SCRealloc(det_ctx->alert_queue, (size_t)(sizeof(PacketAlert) * new_cap));
+        if (unlikely(tmp_queue == NULL)) {
+            WARN_ONCE(SC_ERR_MEM_ALLOC, once_err, "failed to allocate %" PRIuMAX " bytes",
+                    (uintmax_t)(sizeof(PacketAlert) * det_ctx->alert_queue_capacity));
+            /* queue capacity didn't change */
+            return det_ctx->alert_queue_capacity;
+        }
+    }
+    det_ctx->alert_queue = tmp_queue;
+    det_ctx->alert_queue_capacity = new_cap;
+    SCLogNotice("Alert queue size doubled: %u elements, bytes: %" PRIuMAX "",
+            det_ctx->alert_queue_capacity,
+            (uintmax_t)(sizeof(PacketAlert) * det_ctx->alert_queue_capacity));
+    return new_cap;
+}
+
+/**
+ * \brief Append signature to local packet alert queue for later preprocessing
+ */
+void AlertQueueAppend(DetectEngineThreadCtx *det_ctx, const Signature *s, Packet *p, uint64_t tx_id,
+        uint8_t alert_flags)
+{
+    /* first time we see a drop action signature, set that in the packet */
+    /* we do that even before inserting into the queue, so we save it even if appending fails */
+    if (p->alerts.drop.action == 0 && s->action & ACTION_DROP) {
+        p->alerts.drop.num = s->num;
+        p->alerts.drop.action = s->action;
+        p->alerts.drop.s = (Signature *)s;
+        SCLogDebug("Set PacketAlert drop action. s->num %" PRIu32 "", s->num);
+    }
+
+    uint16_t pos = det_ctx->alert_queue_size;
+    if (pos == det_ctx->alert_queue_capacity) {
+        /* we must grow the alert queue */
+        if (pos == AlertQueueExpand(det_ctx)) {
+            /* this means we failed to expand the queue */
+            det_ctx->p->alerts.discarded++;
+            return;
+        }
+    }
+    det_ctx->alert_queue[pos].num = s->num;
+    det_ctx->alert_queue[pos].action = s->action;
+    det_ctx->alert_queue[pos].flags = alert_flags;
+    det_ctx->alert_queue[pos].s = s;
+    /* set the tx_id if the frame has it */
+    det_ctx->alert_queue[pos].tx_id = (tx_id == UINT64_MAX) ? 0 : tx_id;
+    det_ctx->alert_queue[pos].frame_id =
+            (alert_flags & PACKET_ALERT_FLAG_FRAME) ? det_ctx->frame_id : 0;
+
+    SCLogDebug("Appending sid %" PRIu32 ", s->num %" PRIu32 " to alert queue", s->id, s->num);
+    det_ctx->alert_queue_size++;
+    return;
+}
+
+/** \internal
+ * \brief sort helper for sorting alerts by priority
+ *
+ * The Signature::num field is set based on internal priority. Higher priority
+ * rules have lower nums.
+ */
+static int AlertQueueSortHelper(const void *a, const void *b)
+{
+    const PacketAlert *pa0 = a;
+    const PacketAlert *pa1 = b;
+    if (pa1->num == pa0->num)
```

Comment:
Related -- can you update the comment in `decode.h` for the struct `PacketAlert` to reflect what's actually sorted? There, the comments state that only `num` an `action` are used for sorting. I think `tx_id` needs a similar comment.